src/lib/util/view.h

```diff
@@ -0,0 +1,154 @@
+		bool operator>=(const iterator &that) const { return m_ptr >= that.m_ptr; }
+		iterator operator+(size_type i) const { return iterator(m_ptr + i); }
+		iterator operator-(size_type i) const { return iterator(m_ptr - i); }
+		difference_type operator-(const iterator &that) const { return m_ptr - that.m_ptr; }
+
+		iterator& operator++() { m_ptr++; return *this; }
+		iterator operator++(int) { iterator result = *this; m_ptr++; return result; }
+
+		reference operator*() const { return *m_ptr; }
+		pointer operator->() const { return m_ptr; }
+		operator T*() const { return m_ptr; }
+
+	private:
+		T *m_ptr;
+	};
+
+	class const_iterator
+	{
+	public:
+		typedef std::ptrdiff_t difference_type;
+		typedef T value_type;
+		typedef const T &const_reference;
+		typedef const T *const_pointer;
+
+		typedef std::random_access_iterator_tag iterator_category;
+
+		const_iterator(const T *ptr) : m_ptr(ptr) { }
+		const_iterator(const iterator &that) : m_ptr(that) { }
+		const_iterator(const const_iterator &that) : m_ptr(that) { }
+
+		const_iterator& operator=(const const_iterator &that) { m_ptr = that.m_ptr; return *this; }
+		bool operator==(const const_iterator &that) const { return m_ptr == that.m_ptr; }
+		bool operator!=(const const_iterator &that) const { return m_ptr != that.m_ptr; }
+		bool operator<(const const_iterator &that) const { return m_ptr < that.m_ptr; }
+		bool operator>(const const_iterator &that) const { return m_ptr > that.m_ptr; }
+		bool operator<=(const const_iterator &that) const { return m_ptr <= that.m_ptr; }
+		bool operator>=(const const_iterator &that) const { return m_ptr >= that.m_ptr; }
+		const_iterator operator+(size_type i) const { return const_iterator(m_ptr + i); }
+		const_iterator operator-(size_type i) const { return const_iterator(m_ptr - i); }
+		difference_type operator-(const const_iterator &that) const { return m_ptr - that.m_ptr; }
+
+		const_iterator& operator++() { m_ptr++; return *this; }
+		const_iterator operator++(int) { const_iterator result = *this; m_ptr++; return result; }
+
+		const_reference operator*() const { return *m_ptr; }
+		const_pointer operator->() const { return m_ptr; }
+		operator T*() const { return m_ptr; }
+
+	private:
+		const T *m_ptr;
+	};
+
+	view(T *ptr, std::size_t size)
+		: m_begin(ptr)
+		, m_end(ptr + size)
+	{
+	}
+
+	view(const view &that)
+		: m_begin(that.m_begin)
+		, m_end(that.m_end)
+	{
+	}
+
+	// iteration
+	iterator begin() { return m_begin; }
+	const_iterator begin() const { return m_begin; }
+	const_iterator cbegin() const { return m_begin; }
+	iterator end() { return m_end; }
+	const_iterator end() const { return m_end; }
+	const_iterator cend() const { return m_end; }
+
+	// capacity
+	size_type size() const { return m_end - m_begin; }
+	size_type max_size() const { return size(); }
+	bool empty() const { return size() > 0; }
+
+	// element access
+	reference front() { return *this[0]; }
```

Comment:
This is very, very wrong.  What's the precedence of suffix vs prefix operators in C++?  It's effectively `*(this[0])` when you want `(*this)[0]` (you're subscripting the `this` pointer, not the object).  It's clearer to write `return operator[](0)` anyway.